theIDinside/mdebug
# Polish test suite

# Test Suite + Type Management & Value creation

## Test suite
This PR refactors the test suite to automatically pick up (and remove) new tests and tests suites.

Now, to add a new test suite with some sub tests do the following:

- Add new `testFile.js` to `test/driver`
- Add the following code to the file:
```js
async function testSomeFeature(DA) { 
  /* ... test logic, all failing tests should throw exception ...*/
  /* Debug Adapter Client is injected via DA parameter. see client.js for use. */
}

const tests = {
  testSomeFeature: testSomeFeature,
}

module.exports = {
  tests: tests,
}
```
- Re-configure cmake. This will trigger execution of `export_tests.js` which itself will read all .js files in the driver folder and expose the tests given to it in the exported `tests` module variable. The test suite and the sub tests in it can now be reached via `ctest -R testSomeFeature`

## `ValueResolver`, `ValueVisualizer`, `MemoryContentObject` and it's derived types
Added these three base classes which have the responsibility for value creation and display.
ValueResolver has the responsibility of how to construct new values from existing values. So, say we have a `Value` object that represents an array variable in the source program. To construct it's sub objects (the actual elements), there's a `ArrayResolver` that produces those elements. There's also `DefaultStructResolver` (default resolver for structs etc, resolves member variables), and a `CStringResolver` which resolves the contents of a string, that's behind a `const char*` for instance.

Working together with the resolvers are the visualizers. There's a `CStringVisualizer` which works in tandem with it's resolver to ultimately display the value (in a `variables` request). When it comes to `ArrayVisualizer` it's sole job basically becomes printing the type name, since actually displaying values in the `Debug Adapter Protocol` works by requesting a variable by variables reference. There's also a `PrimitiveVisualizer` which writes primitive values (I guess the visualizer, sort of can be called the "pretty printer" of GDB, although full pretty printer functionality is the use of both the resolver and the visualizer).

The backing memory (i.e. the actual contents of a value in the tracee) of a `Value` is found in `MemoryContentObject` which also is a base class. Two derived types exist, an `Eager` and a `Lazy` type, these are self explanatory. The `Lazy` one will read from the tracee at the very last moment (right before the data is actually needed) where as the `Eager` one reads up front. For instance, if we have an array of 200k elements, we might not want to read it all at once. That would be stupid.
